src/tracing/trace_flamegraph.c

```diff
@@ -0,0 +1,304 @@
+	xdebug_hash_add(context->functions, key->d, key->l, function);
+
+	xdebug_str_free(key);
+}
+
+static inline flamegraph_function *fg_function_find(const xdebug_trace_flamegraph_context *context, const int function_nr) {
+	flamegraph_function *function;
+	xdebug_str          *key = fg_function_key(function_nr);
+
+	xdebug_hash_find(context->functions, key->d, key->l, (void*) function);
+
+	xdebug_str_free(key);
+
+	return function;
+}
+
+static inline flamegraph_function *fg_function_delete(const xdebug_trace_flamegraph_context *context, const int function_nr) {
+	flamegraph_function *function;
+	xdebug_str          *key = fg_function_key(function_nr);
+
+	xdebug_hash_delete(context->functions, key->d, key->l);
+
+	xdebug_str_free(key);
+
+	return function;
+}
+
+/* Find parent function in xdebug stack, which is Fiber-safe. */
+static inline function_stack_entry *fg_parent_find() {
+	function_stack_entry *parent_fse;
+	int                   parent_index = XDEBUG_VECTOR_COUNT(XG_BASE(stack)) - 1;
+
+	parent_fse = xdebug_vector_element_get(XG_BASE(stack), parent_index);
+
+	return parent_fse;
+}
+
+static inline int compute_inclusive_value(const xdebug_trace_flamegraph_context *context, const function_stack_entry *fse)
+{
+	int value = 0, current_mem;
+
+	switch (context->mode) {
+		case XDEBUG_TRACE_OPTION_FLAMEGRAPH_MEM:
+			/* We compare with 'memory' because 'prev_memory' is not memory when
+			   starting the function execution, 'memory' is. */
+			current_mem = zend_memory_usage(0);
+			if (current_mem < fse->memory) {
+				/* When memory is below 0, flamegraph generator will error, and
+				   you won't have a good visual.
+				   This happens when garbage collection happened during this
+				   function and freed something it didn't allocate, I guess. */
+				value = 0;
+			} else {
+				value = current_mem - fse->memory;
+			}
+			break;
+
+		case XDEBUG_TRACE_OPTION_FLAMEGRAPH_COST:
+			value = xdebug_get_nanotime() - fse->nanotime;
+			break;
+	}
+
+	return value;
+}
+
+xdebug_trace_flamegraph_context *xdebug_trace_flamegraph_init(char *fname, zend_string *script_filename, long options)
+{
+	xdebug_trace_flamegraph_context *tmp_flamegraph_context;
+	char *used_fname;
+
+	tmp_flamegraph_context = xdmalloc(sizeof(xdebug_trace_flamegraph_context));
+	tmp_flamegraph_context->trace_file = xdebug_trace_open_file(fname, script_filename, options, (char**) &used_fname);
+	tmp_flamegraph_context->trace_filename = used_fname;
+
+	return tmp_flamegraph_context->trace_file ? tmp_flamegraph_context : NULL;
+}
+
+void *xdebug_trace_flamegraph_init_cost(char *fname, zend_string *script_filename, long options)
+{
+	xdebug_trace_flamegraph_context *tmp_flamegraph_context = xdebug_trace_flamegraph_init(fname, script_filename, options);
+
+	if (tmp_flamegraph_context) {
+		tmp_flamegraph_context->mode = XDEBUG_TRACE_OPTION_FLAMEGRAPH_COST;
+		tmp_flamegraph_context->functions = xdebug_hash_alloc(64, (xdebug_hash_dtor_t) fg_function_dtor);
```

Comment:
I chose 64 slots, completely random: I just copy-pasted the size used for fibers stacks hashmap.